proxy.go

```diff
@@ -278,35 +278,47 @@ func (rp *reverseProxy) serveFromCache(s *scope, srw *statResponseWriter, req *h
 	}
 
 	// Await for potential result from concurrent query
-	if userCache.AwaitForConcurrentTransaction(key) {
-		cachedData, err := userCache.Get(key)
-		if err == nil {
-			_ = RespondWithData(srw, cachedData.Data, cachedData.ContentMetadata, cachedData.Ttl, http.StatusOK)
-			cacheHitFromConcurrentQueries.With(labels).Inc()
-			log.Debugf("%s: cache hit after awaiting concurrent query", s)
-			return
+	transaction, err := userCache.AwaitForConcurrentTransaction(key)
+	if err != nil {
+		// log and continue processing
+		log.Errorf("failed to await for concurrent transaction due to: %v", err)
+	} else {
+		if transaction.State.IsCompleted() {
+			cachedData, err := userCache.Get(key)
+			if err == nil {
+				_ = RespondWithData(srw, cachedData.Data, cachedData.ContentMetadata, cachedData.Ttl, http.StatusOK)
+				cacheHitFromConcurrentQueries.With(labels).Inc()
+				log.Debugf("%s: cache hit after awaiting concurrent query", s)
+				return
+			} else {
+				cacheMissFromConcurrentQueries.With(labels).Inc()
+				log.Debugf("%s: cache miss after awaiting concurrent query", s)
+			}
 		} else {
-			cacheMissFromConcurrentQueries.With(labels).Inc()
-			log.Debugf("%s: cache miss after awaiting concurrent query", s)
+			if transaction.State.IsFailed() {
+				err = fmt.Errorf("%s: concurrent query failed", s)
+				respondWith(srw, err, http.StatusInternalServerError)
+				return
+			}
+			// We will assess if elapsed time is bigger than half of the max request timeout.
+			// This precondition halts execution
+			if transaction.ElapsedTime > s.user.maxExecutionTime/2 {
+				err = fmt.Errorf("%s: no result found during grace time period", s)
+				respondWith(srw, err, http.StatusRequestTimeout)
+				return
+			}
 		}
-
 	}
 
 	// The response wasn't found in the cache.
 	// Request it from clickhouse.
 	bufferedRespWriter := cache.NewBufferedResponseWriter(srw)
 
 	// Initialise transaction
-	err = userCache.Register(key)
+	err = userCache.Create(key)
 	if err != nil {
 		log.Errorf("%s: %s; query: %q - failed to register transaction", s, err, q)
 	}
-	defer func() {
-		// Eventually unregister ongoing transaction
-		if err = userCache.Unregister(key); err != nil {
```

Comment:
The fact that you replaced the Unregister() call (that happened at the end of the lifetime of the HTTP call no matter what) with Complete() call will at least increase a lot the memory consumption and it will likely create a memory leak when using the inMemoryTransactionRegistry:
With the unregister() logic, a query is put in the inMemoryTransactionRegistry and is destroyed once the proxy gives an answer to the user.
With the new logic, the Complete() call will just modify the state in the inMemoryTransaction for this query and it will be released only when the pendingEntriesCleaner() call. But it seems that this function is almost never called (every time the config is reload): cf 
https://github.com/ContentSquare/chproxy/blob/fa9522a3a741e1cbcbbd369ae66219c7fbe20ddd/cache/transaction_registry_inmem.go#L23
Other side effect with this logic: if a query was timouting (so with the state Fail), CHProxy will almost never try to redo it (since it will be register for a long time in the registery with a failed state).

One possible fix would be:
- to use a hash instead of the (very verbose) Key of the registered entries to save a lot of memory (but I don't know if it will create some side effects, maybe just hashing the query field in the Key could do it)
- to rework the cleanup logic of the newInMemoryTransactionRegistry function so that the cleanup is called every XX seconds
it's not perfect but should be better than the current behavior.

Reply:
Nb: After a second check, one part is already done. I didn't know that the Stringified version of the Key was already an hash. (Maybe we should rename it to Hash() instead of String() since it's a bit misleading).

Reply:
The eventual cleanup is performed by a background task that you can find [here](https://github.com/ContentSquare/chproxy/blob/fa9522a3a741e1cbcbbd369ae66219c7fbe20ddd/cache/transaction_registry_inmem.go#L98).

The change of unregister was necessary, because we have to tell the difference between `transaction.Complete` and `transaction.Fail`.

Reply:
ok my bad, I didn't see that there was an infinite loop in the function pendingEntriesCleaner(), so this thread is irrelevant